library/std/src/sys/sync/condvar/queue.rs

```diff
@@ -0,0 +1,380 @@
+//! A generic `Condvar` implementation based on thread parking and a lockless
+//! queue of threads.
+//!
+//! Not all platforms provide an efficient `Condvar` implementation: the UNIX
+//! `pthread_condvar_t` needs memory allocation, while SGX doesn't have
+//! synchronization primitives at all. Therefore, we implement our own.
+//!
+//! To do so, we keep a list of the [`Thread`]s waiting on the `Condvar` and
+//! wake them up as needed. Access to the list is controlled by an atomic
+//! counter. To notify a waiter, the counter is incremented. If the counter
+//! was previously zero, the notifying thread has control over the list and
+//! will wake up threads until the number of threads it has woken up equals
+//! the counter value. Therefore, other threads do not need to wait for control
+//! over the list because the controlling thread will take over their notification.
+//!
+//! This counter is embedded into the lower bits of a pointer to the list head.
+//! As that limits the number of in-flight notifications, the counter increments
+//! are saturated to a maximum value ([`ALL`]) that causes all threads to be woken
+//! up, leading to a spurious wakeup for some threads. The API of `Condvar` permits
+//! this however. Timeouts employ the same method to make sure that the current
+//! thread handle is removed from the list.
+//!
+//! The list itself has the same structure as the one used by the queue-based
+//! `RwLock` implementation, see its documentation for more information. This
```

Comment:
Is there any way to share the code for that queue, rather than having duplicate code with the same structure?

Reply:
Yes, but it would be a bit complicated, since the queue-based `RwLock` is used on more platforms. Also, since they aren't accessed concurrently the nodes need atomic pointer here, but for `RwLock` that's not the case (there can be multiple reader threads that traverse the queue).